### PR TITLE
feature\exploratory-data-analisys

### PR DESCRIPTION
**Pull request by jakub1090cn**
## Description

This pull request introduces an Exploratory Data Analysis (EDA) in a Jupyter notebook along with several functions for dataset manipulation. The functionalities included in this PR allow for the creation of data subsets and provide tools to display and analyze information about these subsets. This is essential for better understanding the dataset and preparing it for further modeling.

## Testing

The changes introduced in this pull request were manually tested to ensure they function as expected:

Exploratory Data Analysis (EDA) Notebook:

The Jupyter notebook was executed step-by-step to confirm that all cells run without errors.
Visualizations and data summaries were carefully reviewed to ensure they accurately reflect the underlying data.
Dataset Manipulation Functions:

Functions for creating data subsets were tested by applying them to various segments of the dataset. The resulting subsets were checked to confirm that they contain the expected records.
Functions for displaying information about the subsets were manually verified by inspecting the output to ensure it aligns with the data in the subsets.

## Types of Changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Tests update
- [x] Documentation update
- [x] Changes that do not directly affect the source code or tests

## Checklist
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by skrzypczykt** [comment]
TODO:

* Share unzipped data
* Add counting correct/incorrect images to the function and reeturn also a list of bad images (if any) could return an empty list
* Add typehints
* export all functionalities to functions in data.py, plot.py and utils.py files
* os.makedirs(...,exist_ok=True)
* Add sample bold/not images to the repo - agree on the location with the team, link in the README

**Comment by skrzypczykt** [comment]
TODO: Rename Validation to test
Extract 9% of training data to new `validation.csv`
Uplad new_label to the GDrive

File: src/constants.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import os
+
+DATA_PATH = os.path.join("..", "src", "data", "list_attr_celeba.csv")
+DATASET_PATH = os.path.join("..", "src", "data", "list_eval_partition.csv")
+IMAGES_PATH = os.path.join(
+    "..", "src", "data", "img_align_celeba", "img_align_celeba"
+)
```
Comment (skrzypczykt) [inline src/constants.py]: To nie są dobre miejsca do trzymania danych, bo to jest jakby wewnątrz repo. Polecam, żeby to było w configu jednak - pogadajmy o tym.

File: src/data.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    for filename in os.listdir(directory):
+        file_path = os.path.join(directory, filename)
+        try:
+            img = cv2.imread(file_path)
+
+            if img is None or img.size == 0:
+                empty_or_corrupted.append(filename)
+            else:
+                num_correct += 1
```
Comment (skrzypczykt) [inline src/data.py]: to już nie powinno być w bloku try

File: src/data.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        except Exception:
+            empty_or_corrupted.append(filename)
+
+    return empty_or_corrupted, len(empty_or_corrupted), num_correct
+
+
+def create_sets(path: str) -> None:
```
Comment (skrzypczykt) [inline src/data.py]: `create_data_subsets` ? Co ty na to?

File: src/plot.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import pandas as pd
+import os
+import matplotlib.pyplot as plt
+import cv2
+from typing import List, Dict
+
+
+def display_sample_images(df: pd.DataFrame, dir_path: str) -> None:
+    fig, axes = plt.subplots(1, 2, figsize=(10, 5))
+    labels = {-1: "Not Bald", 1: "Bald"}
```
Comment (skrzypczykt) [inline src/plot.py]: to miało wylądować gdzieś w constach

**Review by skrzypczykt — COMMENTED** [review]
Ogólnie jest bardzo ładnie

File: src/data.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    empty_or_corrupted: List[str] = []
+    num_correct: int = 0
+
+    for filename in os.listdir(directory):
+        file_path = os.path.join(directory, filename)
+        try:
+            img = cv2.imread(file_path)
+
+            if img is None or img.size == 0:
+                empty_or_corrupted.append(filename)
```
Comment (skrzypczykt) [inline src/data.py]: work on this

**Comment by skrzypczykt** [comment]
@jakub1090cn czy ta PRka nie powinna też zawierać wpisu do readme z opisem skąd pochodzą dane i jakieś minimalne informacje o nich?

File: src/data.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import pandas as pd
+
+
+def check_images(directory: str) -> Tuple[List[str], int, int]:
+    empty_or_corrupted: List[str] = []
+    num_correct: int = 0
+
+    for filename in os.listdir(directory):
+        file_path = os.path.join(directory, filename)
+        img = cv2.imread(file_path)
```
Comment (skrzypczykt) [inline src/data.py:32]: Nie do końca rozumiem czemy `try` zniknęło zupełnie

**Review by skrzypczykt — COMMENTED** [review]
.

File: src/data.py (hunk 6 above)
Comment (jakub1090cn) [reply src/data.py:32]: Z tego co doczytałem, jeśli cv2.imread nie otworzy pliku zwróci None, a nie wyjątek

Comment (skrzypczykt) [reply src/data.py:32]: ok - to super

File: src/data.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        if img is None or img.size == 0:
+            empty_or_corrupted.append(filename)
+        else:
+            num_correct += 1
+
+    return empty_or_corrupted, len(empty_or_corrupted), num_correct
```
Comment (raodz) [inline src/data.py]: Suggestion: imo the function would be more clear if it just returned empty_or_corrupted list, the lengh of it and the number of correct images are easy to get from it

File: src/data.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+    This function reads the provided subsets and labels CSV files, merges them based on the `image_id` column,
+    and splits the data into training, validation, and test sets. The partition labels are used to separate
+    the data into training (partition 0) and test (partition 1) sets. The training set is further split
+    into training and validation subsets with 9% of the data allocated to validation. The resulting datasets
+    are saved as CSV files.
+    """
+    subsets = pd.read_csv(subsets_path)
+    labels = pd.read_csv(labels_path)
+    df = pd.merge(subsets, labels, how="inner", on="image_id")
```
Comment (raodz) [inline src/data.py]: Suggestion: df could be created by another function so this one was only for creating subsets

**Review by raodz — COMMENTED** [review]
in general the code is fine :)

File: src/data.py (hunk 7 above)
Comment (skrzypczykt) [reply src/data.py]: agreed

**Comment by skrzypczykt** [comment]
@raodz since you are doing a proper review, you need to decide whether you approve the PR or request changes at this point.

**Review by raodz — CHANGES_REQUESTED** [review]
request for changes in:
check_images() - please change the return and probably change of function's name according to this
create_data_subsets() - please split the function into two: function preparing df and function using df to create subsets

File: src/data.py [hunk 9]
```diff
@@ -38,27 +38,52 @@ def check_sample_images(directory: str) -> Tuple[List[str], int]:
     return empty_or_corrupted, num_correct
 
 
-def create_data_subsets(subsets_path: str, labels_path: str) -> None:
+def prepare_dataframe(subsets_path: str, labels_path: str) -> pd.DataFrame:
```
Comment (raodz) [inline src/data.py]: imo the name of this function could be more specific e.g. prepare_merged_dataframe

**Review by raodz — CHANGES_REQUESTED** [review]
three more things i've just noticed:
- pycache folders are still on this branch (in src, tests and main project dir) - should be deleted as on main
- the name of this branch could be changed since there is not only EDA but also data preparation
- it would be good to add some tests for functions from data.py. maybe can we test at least check_sample_images (on some added files e.g. three files, one of them correct, one corrupted and one empty) and prepare_dataframe?

**Review by raodz — APPROVED** [review]
looks good to me :)

**Pull request merged**